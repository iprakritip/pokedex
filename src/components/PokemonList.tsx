import React, {useEffect, useState} from 'react';
import {useOutletContext} from 'react-router-dom';
import {useGetAllPokemons} from '../api/hooks';
import useDebouncer from '../hooks/useDebouncer';
import Loader from './Loader';
import PokemonInfo from './PokemonInfo';
import Pokemons from './Pokemons';
import PokemonsLoader from './PokemonsLoader';
import ReactPaginate from 'react-paginate';

const PokemonList = () => {
  const {searchInput} = useOutletContext<{
    searchInput: string;
  }>();
  const [displayInfo, setDisplayInfo] = useState<boolean>(false);
  const [clickedPokemonId, setClickedPokemonId] = useState(
    Number(localStorage.getItem('id'))
  );
  const [offset, setOffset] = useState(0);
  const itemsPerPage = 60;
  const endOffset = offset + itemsPerPage;
  const [offsetHolder, setOffsetHolder] = useState(0);

  const togglePokeInfo = () => {
    setDisplayInfo(true);
  };
  const increaseClickedId = () => {
    setClickedPokemonId(clickedPokemonId + 1);
  };
  const decreaseClickedId = () => {
    setClickedPokemonId(clickedPokemonId - 1);
  };
  const closePokeInfo = () => {
    setDisplayInfo(false);
  };
  const changeSelectedPokemonId = (id: number) => {
    setClickedPokemonId(id);
  };

  const debouncedSearch = useDebouncer(searchInput, 500);

  const getData = () => {
    // console.log(offset, 'offset');

    const {data, error, loading, fetchMore} =
      useGetAllPokemons(debouncedSearch);
    return {data, error, loading, fetchMore};
  };

  const {data, error, loading, fetchMore} = getData();
  // console.log(data);

  const currentData = data?.pokemon_v2_pokemon.slice(offset, endOffset);
  const pageCount = Math.ceil(data?.pokemon_v2_pokemon.length / itemsPerPage);
  // console.log(pageCount);

  // console.log(data);
  const handlePageClick = (e: {selected: number}) => {
    const newOffset =
      (e.selected * itemsPerPage) % data.pokemon_v2_pokemon.length;
    setOffset(newOffset);
    setOffsetHolder(newOffset);
  };

  useEffect(() => {
    searchInput !== ''
      ? setOffset(0)
      : offsetHolder > 0
      ? setOffset(offsetHolder)
      : setOffset(0);
  }, [searchInput]);

  if (loading) {
    // console.log('loading...');
    return (
      <div className='h-screen w-screen flex justify-center items-center'>
        <PokemonsLoader />
      </div>
    );
  }
  if (error) return <div>error....</div>;

  // console.log(currentData);

  return (
    <div className='pokemonlist flex flex-col w-screen mt-[9rem] px-[5vw]'>
      {/* <div className='modal bg-black opacity-20 h-full w-full absolute'></div> */}
      <div className='flex'>
        <Pokemons
          data={currentData}
          togglePokeInfo={togglePokeInfo}
          displayInfo={displayInfo}
          changeSelectedPokemonId={changeSelectedPokemonId}
          clickedPokemonId={clickedPokemonId}
          
        />
        <PokemonInfo
          displayInfo={displayInfo}
          clickedPokemonId={clickedPokemonId}
          closePokeInfo={closePokeInfo}
          increaseClickedId={increaseClickedId}
          decreaseClickedId={decreaseClickedId}
          togglePokeInfo={togglePokeInfo}
          changeSelectedPokemonId={changeSelectedPokemonId}
        />
      </div>
      <div className='mt-8'>
        <ReactPaginate
          breakLabel='...'
          nextLabel='next >'
          onPageChange={(e) => handlePageClick(e)}
          pageRangeDisplayed={5}
          pageCount={pageCount}
          previousLabel='< previous'
          renderOnZeroPageCount={null}
          containerClassName={'pagination'}
          activeClassName={'active'}
        />
      </div>
    </div>
  );
};

export default PokemonList;
