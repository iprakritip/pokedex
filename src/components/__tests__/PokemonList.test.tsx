import React from 'react';
import {MockedProvider} from '@apollo/client/testing';
import PokemonList from '../PokemonList';
import '@testing-library/jest-dom';
import {render, screen} from '@testing-library/react';
import {BrowserRouter} from 'react-router-dom';
import {PokemonListMocks, searchMocks} from './mocks/PokemonListMocks';

// const mockedUseContext = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useOutletContext: () => ({searchInput: 'saur'}),
}));
describe('Pokemon List', () => {
  const TestComponent = (
    <MockedProvider mocks={PokemonListMocks}>
      <BrowserRouter>
        <PokemonList />
      </BrowserRouter>
    </MockedProvider>
  );
  const SearchTestComponent = (
    <MockedProvider mocks={searchMocks}>
      <BrowserRouter>
        <PokemonList />
      </BrowserRouter>
    </MockedProvider>
  );
  test('should render pokemons', async () => {
    const {baseElement} = render(TestComponent);
    expect(await screen.findByText(/BULBASAUR/i)).toBeInTheDocument();
    expect(baseElement).toMatchSnapshot();
  });

  test('should display search results', async () => {
    render(SearchTestComponent);
    expect(await screen.findByText(/VENUSAUR-MEGA/i)).toBeInTheDocument();
  });
});
