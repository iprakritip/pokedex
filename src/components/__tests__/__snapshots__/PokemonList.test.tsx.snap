// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon List should render pokemons 1`] = `
<body>
  <div>
    <div
      class="pokemonlist flex w-screen mt-40"
    >
      <div
        class="pokemons flex flex-col items-center gap-28 min-h-screen w-full"
      >
        <div
          class="flex flex-wrap gap-6 justify-between w-full"
        >
          <div
            class="pokemon relative w-[15%] h-[15rem] pt-28 "
          >
            <div
              class="absolute w-full bottom-[45%] z-10 flex justify-center"
            >
              <img
                alt="bulbasaur"
                src="https://projectpokemon.org/images/normal-sprite/bulbasaur.gif"
              />
            </div>
            <div
              class=" w-full bg-white relative border rounded shadow flex flex-col gap-3 items-center pt-8 pb-8"
            >
              <p
                class="text-xs  font-semibold text-gray-500"
              >
                #
                1
              </p>
              <p
                class="font-bold text-gray-800 text-sm"
              >
                BULBASAUR
              </p>
              <div
                class="flex flex-wrap gap-4 justify-center"
              >
                <button
                  class=" bg-grass border-grass false
      false
      false
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      px-3 py-2 border  rounded-md text-xs font-medium text-white"
                >
                  Grass
                </button>
                <button
                  class=" false bg-poison border-poison
      false
      false
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      px-3 py-2 border  rounded-md text-xs font-medium text-white"
                >
                  Poison
                </button>
              </div>
            </div>
          </div>
          <div
            class="pokemon relative w-[15%] h-[15rem] pt-28 "
          >
            <div
              class="absolute w-full bottom-[45%] z-10 flex justify-center"
            >
              <img
                alt="ivysaur"
                src="https://projectpokemon.org/images/normal-sprite/ivysaur.gif"
              />
            </div>
            <div
              class=" w-full bg-white relative border rounded shadow flex flex-col gap-3 items-center pt-8 pb-8"
            >
              <p
                class="text-xs  font-semibold text-gray-500"
              >
                #
                2
              </p>
              <p
                class="font-bold text-gray-800 text-sm"
              >
                IVYSAUR
              </p>
              <div
                class="flex flex-wrap gap-4 justify-center"
              >
                <button
                  class=" bg-grass border-grass false
      false
      false
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      px-3 py-2 border  rounded-md text-xs font-medium text-white"
                >
                  Grass
                </button>
                <button
                  class=" false bg-poison border-poison
      false
      false
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      px-3 py-2 border  rounded-md text-xs font-medium text-white"
                >
                  Poison
                </button>
              </div>
            </div>
          </div>
          <div
            class="pokemon relative w-[15%] h-[15rem] pt-28 "
          >
            <div
              class="absolute w-full bottom-[45%] z-10 flex justify-center"
            >
              <img
                alt="venusaur"
                src="https://projectpokemon.org/images/normal-sprite/venusaur.gif"
              />
            </div>
            <div
              class=" w-full bg-white relative border rounded shadow flex flex-col gap-3 items-center pt-8 pb-8"
            >
              <p
                class="text-xs  font-semibold text-gray-500"
              >
                #
                3
              </p>
              <p
                class="font-bold text-gray-800 text-sm"
              >
                VENUSAUR
              </p>
              <div
                class="flex flex-wrap gap-4 justify-center"
              >
                <button
                  class=" bg-grass border-grass false
      false
      false
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      px-3 py-2 border  rounded-md text-xs font-medium text-white"
                >
                  Grass
                </button>
                <button
                  class=" false bg-poison border-poison
      false
      false
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      px-3 py-2 border  rounded-md text-xs font-medium text-white"
                >
                  Poison
                </button>
              </div>
            </div>
          </div>
          <div
            class="pokemon relative w-[15%] h-[15rem] pt-28 "
          >
            <div
              class="absolute w-full bottom-[45%] z-10 flex justify-center"
            >
              <img
                alt="venusaur-mega"
                src="https://projectpokemon.org/images/normal-sprite/venusaur_mega.gif"
              />
            </div>
            <div
              class=" w-full bg-white relative border rounded shadow flex flex-col gap-3 items-center pt-8 pb-8"
            >
              <p
                class="text-xs  font-semibold text-gray-500"
              >
                #
                10033
              </p>
              <p
                class="font-bold text-gray-800 text-sm"
              >
                VENUSAUR-MEGA
              </p>
              <div
                class="flex flex-wrap gap-4 justify-center"
              >
                <button
                  class=" bg-grass border-grass false
      false
      false
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      px-3 py-2 border  rounded-md text-xs font-medium text-white"
                >
                  Grass
                </button>
                <button
                  class=" false bg-poison border-poison
      false
      false
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      px-3 py-2 border  rounded-md text-xs font-medium text-white"
                >
                  Poison
                </button>
              </div>
            </div>
          </div>
          <div
            class="pokemon relative w-[15%] h-[15rem] pt-28 "
          >
            <div
              class="absolute w-full bottom-[45%] z-10 flex justify-center"
            >
              <img
                alt="venusaur-gmax"
                src="https://projectpokemon.org/images/normal-sprite/venusaur_gmax.gif"
              />
            </div>
            <div
              class=" w-full bg-white relative border rounded shadow flex flex-col gap-3 items-center pt-8 pb-8"
            >
              <p
                class="text-xs  font-semibold text-gray-500"
              >
                #
                10195
              </p>
              <p
                class="font-bold text-gray-800 text-sm"
              >
                VENUSAUR-GMAX
              </p>
              <div
                class="flex flex-wrap gap-4 justify-center"
              >
                <button
                  class=" bg-grass border-grass false
      false
      false
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      px-3 py-2 border  rounded-md text-xs font-medium text-white"
                >
                  Grass
                </button>
                <button
                  class=" false bg-poison border-poison
      false
      false
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      false 
      px-3 py-2 border  rounded-md text-xs font-medium text-white"
                >
                  Poison
                </button>
              </div>
            </div>
          </div>
        </div>
        <button
          class="bg-pink-600 border border-pink-600 text-white w-[14%] text-sm px-6 py-2 rounded-lg"
        >
          Load more
        </button>
      </div>
      <div
        class="h-[75vh] w-96 px-4 pb-4  bg-white shadow border rounded-lg fixed top-40  right-[-25rem] flex flex-col items-center gap-2 transition-position duration-400 z-20 ease-in-out"
      >
        <div
          class="flex flex-col items-center gap-2"
        >
          <div
            class="w-full flex justify-end"
          >
            <a
              href="/"
            >
              <button
                class="text-xs text-gray-500 underline"
              >
                close&gt;&gt;
              </button>
            </a>
          </div>
          <div
            class="w-28"
          >
            <img
              alt="bulbasaur"
              class=""
              src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/home/undefined.png"
            />
          </div>
          <p
            class="text-sm  font-semibold text-gray-500"
          >
            #
          </p>
          <p
            class="font-bold text-lg text-gray-800"
          />
          <div
            class="flex flex-wrap gap-4"
          />
          <div
            class="flex flex-col gap-6 items-center"
          >
            <div
              class="flex flex-col items-center gap-1"
            >
              <p
                class="font-semibold text-sm"
              >
                POKÉDEX ENTRY
              </p>
              <p
                class="text-sm font-medium text-gray-800 text-center"
              />
            </div>
            <div
              class="flex gap-4 justify-center"
            >
              <div
                class="flex flex-col items-center gap-1"
              >
                <p
                  class="font-semibold text-sm"
                >
                  HEIGHT
                </p>
                <p
                  class="bg-slate-100 w-32 py-2 border text-sm border-slate-100 rounded-full text-center"
                >
                   meters
                </p>
              </div>
              <div
                class="flex flex-col items-center gap-1"
              >
                <p
                  class="font-semibold text-sm"
                >
                  WEIGHT
                </p>
                <p
                  class="bg-slate-100 w-32 py-2 border text-sm border-slate-100 rounded-full text-center"
                >
                   kgs
                </p>
              </div>
            </div>
            <div
              class="w-full"
            >
              <div
                class="flex gap-4 justify-between"
              >
                <button
                  class="w-full flex justify-center items-center font-semibold text-sm gap-2 bg-slate-100 border py-2 border-slate-100 rounded-md"
                >
                  Prev
                </button>
                <button
                  class="w-full flex justify-center items-center font-semibold text-sm gap-2 bg-slate-100 border py-2 border-slate-100 rounded-md"
                >
                  Next
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
